reference/solver/cb_gmres_kernels.cpp

```diff
@@ -0,0 +1,460 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/solver/cb_gmres_kernels.hpp"
+
+
+#include <type_traits>
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/stop/stopping_status.hpp>
+
+
+#include "core/solver/cb_gmres_accessor.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The CB_GMRES solver namespace.
+ *
+ * @ingroup cb_gmres
+ */
+namespace cb_gmres {
+
+
+namespace {
+
+
+template <typename ValueType, typename Accessor3d>
+void finish_arnoldi_CGS2(matrix::Dense<ValueType> *next_krylov_basis,
+                         Accessor3d krylov_bases,
+                         matrix::Dense<ValueType> *hessenberg_iter,
+                         matrix::Dense<ValueType> *buffer_iter,
+                         matrix::Dense<remove_complex<ValueType>> *arnoldi_norm,
+                         size_type iter, const stopping_status *stop_status)
+{
+    static_assert(
+        std::is_same<ValueType,
+                     typename Accessor3d::accessor::arithmetic_type>::value,
+        "ValueType must match arithmetic_type of accessor!");
+    constexpr bool has_scalar =
+        gko::cb_gmres::detail::has_3d_scaled_accessor<Accessor3d>::value;
+    using rc_vtype = remove_complex<ValueType>;
+    const rc_vtype eta = 1.0 / sqrt(2.0);
+
+    for (size_type i = 0; i < next_krylov_basis->get_size()[1]; ++i) {
+        arnoldi_norm->at(0, i) = zero<rc_vtype>();
+        for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+            arnoldi_norm->at(0, i) += squared_norm(next_krylov_basis->at(j, i));
+        }
+        arnoldi_norm->at(0, i) = eta * sqrt(arnoldi_norm->at(0, i));
+        // arnoldi_norm->at(0, i) = norm(next_krylov_basis)
+        if (stop_status[i].has_stopped()) {
+            continue;
+        }
+        for (size_type k = 0; k < iter + 1; ++k) {
+            hessenberg_iter->at(k, i) = zero<ValueType>();
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                hessenberg_iter->at(k, i) +=
+                    next_krylov_basis->at(j, i) * krylov_bases(k, j, i);
+            }
+        }
+        // for i in 1:iter
+        //     hessenberg(iter, i) = next_krylov_basis' * krylov_bases(:, i)
+        // end
+        for (size_type k = 0; k < iter + 1; ++k) {
+            for (size_type j = 0; j < next_krylov_basis->get_size()[0]; ++j) {
+                next_krylov_basis->at(j, i) -=
+                    hessenberg_iter->at(k, i) * krylov_bases(k, j, i);
+            }
```

Comment:
The problem is that even for CUDA and HIP, all these functions are unique, and we rarely re-use them. I decided to keep it as it is because understanding the for-loops is simple because each one is quite short. I think it would be more complicated to understand this `finish_arnoldi_CGS` method if it was littered with multiple function calls, where you would need to jump around to get the bigger picture.